### PR TITLE
Misc cleanup & minor extensions

### PR DESCRIPTION
**Pull request by dweindl**
* replace magic numbers and defines by enums
* hdf5 IO& error handling
* remove redundant userdata fields
* ...

### CODE REVIEW

File: src/amici.cpp [hunk 1]
```diff
@@ -214,9 +214,9 @@ UserData *setupUserData(const mxArray *prhs[]) {
 
     readOptionData(z2event)
     readOptionData(qpositivex)
-    readOptionScalar(sensi,int)
+    readOptionScalar(sensi,AMI_sensi_order)
```
Comment (FFroehlich) [inline src/amici.cpp]: This needs to be handled differently now. For both AMI_SENSI_ORDER_SECOND and AMI_SENSI_ORDER_SECOND_DIRECTIONAL we will have a value sensi==2. So we need to change this value depending on the mode with which the model was augmented (this is currently realised via the AMI_o2mode_TAG which is initialised in the init_modeldims function)

**Comment by FFroehlich** [comment]
I think with respect to the AMI_sensi_order_TAG, I would not include one enum for directional and full because this will be specified during compile time (hardcoded in init_modeldims) while the sensi_order flag is set at runtime (it would not make sense to specify directional when you only compiled for second order).

File: src/amici.cpp (hunk 1 above)
Comment (dweindl) [inline src/amici.cpp]: Ah, great. More intuitive. I will remove `AMI_SENSI_ORDER_SECOND_DIRECTIONAL` then.

Comment (dweindl) [inline src/amici.cpp]: Should be fixed in f37a398.

**Pull request merged**